### PR TITLE
Add gitlab config

### PR DESCRIPTION
**Pull request by TriplEight**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -215,6 +219,32 @@ script:
...
+### grcov with GitLab CI and codecov.io
+
+```yaml
+codecov:
+  stage:                           coverage
+  variables:
+    CARGO_INCREMENTAL:             0
+    RUSTFLAGS:                     "-Zprofile -Zmir-opt-level=0 -Ccodegen-units=1
+                                      -Copt-level=0 -Clink-dead-code -Coverflow-checks=off"
+    # We removed the `-Cinline-threshold=0` flag from the above `RUSTFLAGS` since it was bugged
```
Comment (marco-c) [inline README.md:231]: This is no longer in the README, since 2d37c3aa6d9ddca95fd95e695fc925bfb428745d

File: README.md [hunk 2]
```diff
@@ -215,6 +219,32 @@ script:
...
+    # at the time and lead to inlining of functions that shouldn't be inlined for the coverage
+    # report. (More information here: https://github.com/Kogia-sima/rust-covfix/issues/2)
+  script:
+    # RUSTFLAGS are the cause target cache can't be used here
+    - unset "CARGO_TARGET_DIR" # only needed if it's not default
+    - cargo clean
```
Comment (marco-c) [inline README.md:237]: Why the initial clean?

File: README.md [hunk 3]
```diff
@@ -215,6 +219,32 @@ script:
...
 ```
 
+### grcov with GitLab CI and codecov.io
+
+```yaml
+codecov:
+  stage:                           coverage
+  variables:
+    CARGO_INCREMENTAL:             0
+    RUSTFLAGS:                     "-Zprofile -Zmir-opt-level=0 -Ccodegen-units=1
```
Comment (marco-c) [inline README.md:229]: `-Zmir-opt-level` is not in the list above, why is it useful? If it is, we should add it everywhere

File: README.md [hunk 4]
```diff
@@ -215,6 +219,32 @@ script:
...
+                                      -Copt-level=0 -Clink-dead-code -Coverflow-checks=off"
+    # We removed the `-Cinline-threshold=0` flag from the above `RUSTFLAGS` since it was bugged
+    # at the time and lead to inlining of functions that shouldn't be inlined for the coverage
+    # report. (More information here: https://github.com/Kogia-sima/rust-covfix/issues/2)
+  script:
+    # RUSTFLAGS are the cause target cache can't be used here
+    - unset "CARGO_TARGET_DIR" # only needed if it's not default
+    - cargo clean
+    - cargo test --verbose --all-features --no-fail-fast --workspace
```
Comment (marco-c) [inline README.md:238]: Isn't `--workspace` useless here?

File: README.md [hunk 5]
```diff
@@ -215,6 +219,32 @@ script:
...
+    - cargo build --verbose --all-features --workspace
```
Comment (marco-c) [inline README.md:239]: Let's build first, and then run the tests

File: README.md [hunk 6]
```diff
@@ -215,6 +219,32 @@ script:
...
+    # coverage with branches
+    - grcov ./target -s . -t lcov --llvm --branch --ignore-not-existing --ignore "/*" --ignore "tests/*" -o lcov-w-branch.info
+    - rust-covfix -o lcov-w-branch_correct.info lcov-w-branch.info
+    # We'd like to not use a remote bash script for uploading the coverage reports,
```
Comment (marco-c) [inline README.md:243]: Let's remove this comment

File: README.md [hunk 7]
```diff
@@ -215,6 +219,32 @@ script:
...
+    # report. (More information here: https://github.com/Kogia-sima/rust-covfix/issues/2)
+  script:
+    # RUSTFLAGS are the cause target cache can't be used here
+    - unset "CARGO_TARGET_DIR" # only needed if it's not default
+    - cargo clean
+    - cargo test --verbose --all-features --no-fail-fast --workspace
+    - cargo build --verbose --all-features --workspace
+    # coverage with branches
+    - grcov ./target -s . -t lcov --llvm --branch --ignore-not-existing --ignore "/*" --ignore "tests/*" -o lcov-w-branch.info
+    - rust-covfix -o lcov-w-branch_correct.info lcov-w-branch.info
```
Comment (marco-c) [inline README.md:242]: Let's not use rust-covfix, if we want to use it, it should be run automatically by grcov

File: README.md [hunk 8]
```diff
@@ -12,18 +12,19 @@ Linux, OSX and Windows are supported.
...
+
 * [Usage](#usage)
-    * [LCOV output](#lcov-output)
-    * [Coveralls/Codecov output](#coverallscodecov-output)
-    * [grcov with Travis](#grcov-with-travis)
-    * [Auto Formatting](#auto-formatting)
+  * [LCOV output](#lcov-output)
+  * [Coveralls/Codecov output](#coverallscodecov-output)
+  * [grcov with Travis](#grcov-with-travis)
+  * [Auto Formatting](#auto-formatting)
```
Comment (marco-c) [inline README.md:20]: Could you revert the changes that are unrelated to this PR?

File: README.md (hunk 2 above)
Comment (TriplEight) [reply README.md:237]: it's in case if there's caching set up, to make sure the cache is not used for this job. Same for `unset "CARGO_TARGET_DIR"`

File: README.md (hunk 1 above)
Comment (TriplEight) [reply README.md:231]: removed

File: README.md (hunk 5 above)
Comment (TriplEight) [reply README.md:239]: not sure why the difference in this case, but ok.

File: README.md (hunk 4 above)
Comment (TriplEight) [reply README.md:238]: it's useful for the [project](https://github.com/paritytech/ink/blob/master/.gitlab-ci.yml#L114) I've implemented it for, which is a large virtual workspace

File: README.md (hunk 3 above)
Comment (TriplEight) [reply README.md:229]: mostly for no-landing-pads. Tests have shown that with it, thee coverage is better.

File: README.md (hunk 7 above)
Comment (TriplEight) [reply README.md:242]: Without it, grcov looses 10-20% accuracy, it's a really good thing. Can it be ran by grcov?

File: README.md (hunk 8 above)
Comment (TriplEight) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (TriplEight) [reply README.md:225]: however, it requires nightly

**Comment by TriplEight** [comment]
@marco-c anything else?

**Comment by marco-c** [comment]
Sorry for the delay @TriplEight, I've been pretty busy with other things. I've recently added support for source-based coverage (in version 0.6.0). It'd be nice to show both source-based coverage and gcov-based coverage in the gitlab config (like we are doing for travis right now).

**Comment by TriplEight** [comment]
Ok, will do. Wanted to check out that source-based thing anyway.

**Comment by marco-c** [comment]
@TriplEight ping

**Comment by marco-c** [comment]
Closing for lack of activity. Feel free to reopen if you still intend to work on this.
I've filed #814 to track the doc improvement.
